### DISCUSSION TITLE
CM3 has nothing to play!

### DISCUSSION THREAD

**Discussion by nsgnkhibdk2cls0f** [modula3/cm3#1136] [General]
We are not developing something from scratch, like an operating system so we need to use external libraries. Good language is only one aspect, we need good libraries too. CM3 has no bindings to useful external C libraries. libxml2? Not available! libcurl? Not available! gtk? Not available! sdl2? Not available! glfw? Also not available! Windows API? An outdated version created before I was even born and not updated since then!

Ever wonder why no one used your language other than you? Now you have the answer!

Look at the number of C libraries of FreeBasic and be shamed!

https://github.com/freebasic/fbc/tree/master/inc

Have a look at this, too: https://github.com/edubart/nelua-decl

Even toy languages have users, because they included the needed batteries! But you are not!

**Comment by ghost** [comment]
There is https://github.com/demoitem/m3swig and https://github.com/demoitem/swigm3. Don't know which to use.

**Comment by ghost** [comment]
CM3 targets the same market as Go but lacked the polished of Go. Documentation for CM3 is bad, the language manual is organized a way that more difficult for you to navigate and find what you needed. It also lacked many things, it's not a complete reference you could rely on to learn the language. You have to buy books, various Modula-3 books. These books most of the time is duplicated of each others but they do provide information not available on the language manual.

**Reply by ghost** [reply]
One example of how the language manual make it difficult for you to navigate and find the things you needed. I want to find more information about they keyword EXTERNAL (this is used for C libraries). I have real difficult finding it. Switching back and forth between sections of this document, each section is very short and lacked of needed information:

http://modula3.github.io/cm3/reference/complete/html/Modula_3_Language_definitio.html

Oh, finally I found it: http://modula3.github.io/cm3/reference/complete/html/2_8_5Pragmas.html

But this document is useless, it doesn't tell me how to use EXTERNAL at all.

**Reply by ghost** [reply]
You have to ask questions, various questions, even for the most simple ones that is so easy to see from the tutorials of other languages, ask people here that really used CM3 and take part in development of CM3. Various people come before you and me encountered these hassles and asked and was answered. Unfortunately, they are flagged by Github. These precious resources now all vaporized, like the people asked them.

**Reply by ghost** [reply]
If the people here really cared about you, they should write tutorials incorporated these things and update their documentation so new people don't have to ask about the very same problems again and again, but they don't do this!

**Reply by ghost** [reply]
Instead they whined about how their language is unpopular and no one use it! They blamed Debian and the FSF. Playing victim is much easier to do than doing something real that's useful!

**Reply by mikanystrom** [reply]
<*EXTERNAL*> is not part of the language definition itself.  It is a pragma that tells the linker that the implementation of something is provided externally.  You can find examples of it in various places in the cm3 repo.  Look at the coroutines for an example.

https://github.com/modula3/cm3/tree/master/m3-libs/m3core/src/coroutine/UCONTEXT

Look at the relationship between ContextC.c and ContextC.i3 and you'll see how the C routines are given a Modula-3 interface using the <*EXTERNAL*> pragma.  That's all it is.

You're right that Modula-3 doesn't have much in the way of tutorials.  But when you see a gap that annoys you with free software.. if it really annoys you.. feel *free* to plug it!  That's the point, isn't it?

I know my name seems to have "Maintainer" next to it for CM3.  I am not quite sure who put that there.  I'm just a user.  I'm actually very happy with the system.  I don't know who you're talking about that's "whining" how the language is unpopular and no one uses it.  For myself, I am not sure it is entirely a bad thing that it is unpopular, and I definitely use it---almost every day.  It's not something I tend to whine about.  I gave you the explanations about Debian and FSF to try to answer some of your (I thought quite reasonable, by the way) observations about the current state of affairs.  I didn't mean to appear to put blame for anything in particular on Debian and FSF: they have their own goals, and I think it's understandable that they act the way they do.  They don't owe the Modula-3 language or language users' community anything.

Important note: there is no "board of directors of CM3" or anything like that.  All I said was simply my own personal opinions and observations.  If you want to help, please go ahead.

I do not want to speak for anyone else, but I get the feeling that most others that are involved with cm3 are people like me: people who use it as users, and don't mind contributing whatever we can to improve it, but also people for whom it's not in any of our job description that we are to spend time evangelizing or popularizing Modula-3 or CM3, or even improving it in any particular way.  We're also not a paid help desk, but obviously people do generally want to be helpful and answer questions, so feel free to ask whatever you want.  But maybe tone down the judgments a bit...

**Comment by ghost** [comment]
Let me tell you REAL data type is float on C and LONG REAL is double on C. The problem is, the built-in library to deal with Math on CM3 are all about REAL, they required REAL input and returned REAL output. To be able to process LONG REAL, you have to delved into the C library!

**Reply by ghost** [reply]
I told them to introduce compiler directives or compiler options to workaround this: https://github.com/modula3/cm3/issues/1120

They refused. In fact their PRAGMA statement is very useless, unlike on C:

http://modula3.github.io/cm3/reference/complete/html/2_8_5Pragmas.html

(Again, very difficult to find, as it's expected to be put in Statements section, but instead it's on Syntax sections, as I told you, the language manual is a mess, who organized it must be a fool or too lazy or careless).

**Reply by mikanystrom** [reply]
I.

If you are just USING Modula-3, you should probably be using Math.i3 for your numerical work

https://github.com/modula3/cm3/blob/master/m3-libs/libm3/src/arith/POSIX/Math.i3

everything in there is LONGREAL (same as C---because it's really just an EXTERNAL linkage to libm on your platform).

II.

As for your comment re: compiler directives or compiler options, you didn't give a single concrete example of something you wanted to see changed in the language.  Just said that we shouldn't use quake.  OK I can see that... maybe it would be nice to have a version of cm3 that worked like a normal compiler, for some uses.  

But I'll just stick with quake because it works better than make could for Modula-3...

I also don't understand how this is a "reply" to your comment re: REAL vs LONGREAL.  Are you saying that you want a compiler option to change the size of REAL?

**Comment by ghost** [comment]
It's no where in the documentation for you to know. Only if you have read many submitted issues and discussions like me you could know that, if you use MinGW on Windows, CM3 is unable to produce dynamic library (aka DLL), DLL is only possible with MSVC compiler!

**Reply by ghost** [reply]
The way they make DLL is also very foolish, see: https://github.com/modula3/cm3/tree/master/examples/win32-dll/src

You will need to write C code. Other languages could make DLL without a single line of C code!

**Reply by ghost** [reply]
This is about dynamic library, about static library (.lib of MSVC and .a of MinGW), there is completely no information. It's not a surprise for me if they told you CM3 is not possible to make static library too!

**Comment by RodneyBates** [comment]
In cm3/m3-libs/m3core/src/float/IEEE/m, you will find:

ExtendedFloat.m3
Extended.i3
LongFloat.m3
LongReal.i3
m3makefile
RealFloat.m3
Real.i3

The other sizes of floats right beside the Real libraries.

If IEEE is not your float representation, there are several others in
directories right beside IEEE.

**Comment by RodneyBates** [comment]
On 1/7/23 01:54, tn1997tn wrote:
>
> CM3 targets the same market as Go but lacked the polished of Go. Documentation for CM3 is bad, the language manual is organized a way that more difficult for you to navigate and find what you needed. It also lacked many things, it's not a complete reference you could rely on to learn the language.
>
I believe it is very complete, perhaps unusually so.
Does the compactness of the language make is seem
incomplete?  (Others have thought so, incorrectly.)

What can you cite that is lacking in the language definition?

**Reply by ghost** [reply]
REVEAL keyword. It's not documented. People has to learn why there is the REVEAL keyword in books.

**Reply by ghost** [reply]
This is only one example. Another is EXTERNAL, it only very briefly mentioned, no further explanation about how to use it or example code.

**Reply by ghost** [reply]
Too many similar examples. It's you that ignored them.

**Reply by ghost** [reply]
The language definition is not compact but careless work. Missed so many of information. Yeah, the language definition is not a manual. But you have no other documents available (for free), it's the only source people could learn about the language. If you think the language definition should be compact, then please provide the manual or tutorials, can you?

**Reply by ghost** [reply]
I can learn other languages for free, completely free using the resource freely available by the internet. Can I do the same with CM3? No. You will need books. Many books. That teach only basic knowledge. Not something worth to buy like a book about the advanced concept of C++ (that you can't see on free only tutorial sites).

**Reply by ghost** [reply]
Be ashamed!

https://www.tutorialspoint.com/index.htm

https://www.javatpoint.com

https://www.geeksforgeeks.org

**Reply by mikanystrom** [reply]
REVEAL is very much documented!  It has its own page.  Called... "revelations".  Seems sensible enough to me?

https://www.cs.purdue.edu/homes/hosking/m3/reference/revelations.html

That being said, it's definitely true that you're not going to learn the *intent* of REVEAL from the language definition.

You don't need to buy books to "get books" on Modula-3.

If you're really interested, read the SRC research reports.  They are all available, for free, from HP, in soft copy.

This one is a must and has a very tutorial nature:

https://www.hpl.hp.com/techreports/Compaq-DEC/SRC-RR-113.pdf

the threads one is good

https://www.hpl.hp.com/techreports/Compaq-DEC/SRC-RR-35.pdf

Network objects

https://hplabs.itcs.hp.com/techreports/Compaq-DEC/SRC-RR-115.html

Implementation of Rd and Wr (lots of examples of REVEAL in there!---if you read & understand this report you really shouldn't have any more questions about REVEAL):

https://www.hpl.hp.com/techreports/Compaq-DEC/SRC-RR-53.html


You can find a whole listing of the Compaq-DEC reports here

https://www.hpl.hp.com/hplabs/index/Compaq-DEC

(many are about other topics)

Sure some of this stuff is a bit difficult to find, but it is out there.  It is a bit unfair to say that there's no tutorial information available.

I've suggested Harbison's book because it's a good way "to get started".  But if you want to go deeper, the SRC reports are absolutely full of information.

Another very important resource, the best really, is to READ the code that was written at SRC.  Most of it is of (I think) very high quality and really shows how to use most of the language features well.  For example, study the generic modules for Table.

https://github.com/modula3/cm3/blob/master/m3-libs/libm3/src/table/Table.ig

https://github.com/modula3/cm3/blob/master/m3-libs/libm3/src/table/Table.mg

it has generics, REVEAL, lots of documentation, use of exported procedures vs. non-exported procedures, etc.  It is of high tutorial value.  And there are many similar examples in the repo.  Go to town!

**Reply by mikanystrom** [reply]
oh and you should look at Luca Cardelli's paper, also:

http://lucacardelli.name/papers/modula3typesystem.a4.pdf

**Comment by RodneyBates** [comment]
On 1/7/23 02:17, tn1997tn wrote:
>
> It's no where in the documentation for you to know. Only if you have read many submitted issues and discussions like me you could know that, if you use MinGW on Windows, CM3 is unable to produce dynamic library (aka DLL), DLL is only possible with MSVC compiler!
>

I don't know about Windows DLLs, but on Linux and similar
systems, libraries will be compiled as .so (dynamic) by
default.  Static are also possible, but off hand, I don't
remember how you ask.

Jay?

**Reply by ghost** [reply]
How? How to make a static library? Can you point my to something that work?

**Reply by ghost** [reply]
I want to see something real I could build and run, not what you think it should be.

**Reply by ghost** [reply]
Confirmed I could make static library on Windows. I have never tried to make one, but I found them on the CM3 distribution. Even though I'm using MinGW, the extension for static library is still .lib like what MSVC named them rather than .a extension used by MinGW.

**Reply by ghost** [reply]
About DLLs, I don't see any DLLs on my CM3 distribution!

**Comment by VictorMiasnikov** [comment]
( I can not write a "reply" on smartphone...
It is Github engine disadvantage.
I will repost this "comment" as "reply" )

 M3swig  is more modern version of "SWIG for Modula-3" 

 Variant with some technical challenges can be found by this URL
https://github.com/demoitem/m3swig/pull/4

**Reply by ghost** [reply]
This is not technical challenges. It's the carelessness of the developer to keep these leftovers. He expects people's machine to be the same as him?

**Reply by ghost** [reply]
Seems this is no longer true, swigm3 is what he is currently working on, at least it seems to be so:

https://github.com/demoitem/swigm3

Notice he synced his outdated repo with upstream swig? I guess he will work on this repo and even attempt to upstream Modula-3 support.

**Reply by nsgnkhibdk2cls0f** [reply]
I hope you could have your fork synced with upstream swig so it will be easier to create interface to useful C libraries. FreeBasic alone has about three different tools to translate C headers to FreeBasic headers and a lot of passionate users good at C to do the job manually where the translator fails. It will be the first step to solve the problem presented on this thread. Note: this thread is not about whining how bad the language or how bad the documentation is!

**Comment by VictorMiasnikov** [comment]
( It is future reply "number 2"...)

}  if you use MinGW on Windows, CM3 is unable to produce dynamic library (aka DLL), DLL is only possible with MSVC compiler!

 It looks like that I personally write about this.
And now I should comment this news.

ToDo

**Comment by VictorMiasnikov** [comment]
( There is another surprise: I can't edit own messages on smartphone...)

Cm3 AMD64_MINGW was born at 2021 year.
Cm3 AMD64_NT was born at 2013 year.

In case when AMD64_MINGW is not perfect then using AMD64_NT is "good choice".

 I _special_ write some month ago about one disadvantage of _current_ state of "cm3 AMD64_MINGW" because _today_ using of AMD64_NT continue be "a best choice" on OS Windows

**Reply by ghost** [reply]
You don't care about free software world so the free software world don't care about you! Yes, MS made Windows and MS also made MSVC, so MSVC is the best on Windows, no one will argue with this. But why people spend their precious time with something like MinGW? They even ported DDK so writing drivers in MinGW become possible. Ever wonder why? Of course it's out of the mindset of you so you can't understand!

**Comment by VictorMiasnikov** [comment]
= =

The way they make DLL is also very foolish, see: https://github.com/modula3/cm3/tree/master/examples/win32-dll/src

You will need to write C code. Other languages could make DLL without a single line of C code!

= =

IMHO, this source code example is dated 2000 - 2009 years?

 It looks like, it correlated with very early state of "cm3 I386_NT".

 Today we have  10 .. 90 .dll ( or .so on Unix  / Linux) inside cm3-all- * distributive(s).

**Reply by ghost** [reply]
So DLL not work with MinGW is outdated information?

**Reply by ghost** [reply]
About i386, i386 MinGW is not working. https://github.com/modula3/cm3/issues/1115

**Reply by ghost** [reply]
I don't see any DLLs on my MinGW based CM3 distribution! You provided wrong information again!

**Comment by VictorMiasnikov** [comment]
= =
Let me tell you REAL data type is float on C and LONG REAL is double on C. The problem is, the built-in library to deal with Math on CM3 are all about REAL, they required REAL input and returned REAL output. To be able to process LONG REAL, you have to delved into the C library!
= =

 IMHO, inside cm3 ecosystem all good with "64 bit reals" a-ka double.

 One of most biggest library is "algebra" .dll / .so
All this lib based on double precision.

**Reply by ghost** [reply]
As I said, you have to delve into the C library to work with LONG REAL because CM3 libraries expect REAL.

**Reply by ghost** [reply]
https://github.com/modula3/cm3/issues/1079

**Reply by ghost** [reply]
https://github.com/modula3/cm3/issues/1076

**Comment by thielema** [comment]
On Sun, 8 Jan 2023, tn1997tn wrote:

> REVEAL keyword. It's not documented. People has to learn why there is the REVEAL keyword in books.

http://modula3.github.io/cm3/reference/revelations.html

**Reply by ghost** [reply]
Only to show your language definition is organized in a way that more difficult for people to find things. When I expect to find it on one section it's indeed on another section that has nothing to do with it. e.g: PRAGMA, fragma is a statement, it should be on Statements, why it's on Syntax?

**Reply by ghost** [reply]
Could you find where in the language definition talk about BRANDED OBJECT? Please tell me. I can't find it! If your language definition is a plain text I at least could use grep to find the thing I need. It's in HTML to help navigation between sections, but the way it organized the thing is just make it harder to find the thing you need.

**Reply by ghost** [reply]
Without this question I have no idea about it: https://github.com/modula3/cm3/issues/1075

**Reply by mikanystrom** [reply]
Pragmas are not statements.  Pragmas are compiler directives.

https://en.wikipedia.org/wiki/Statement_(computer_science)

**Reply by mikanystrom** [reply]
BRANDED is where I would expect it, under reference types.

https://www.cs.purdue.edu/homes/hosking/m3/reference/refs.html

To wit:

In both the traced and untraced cases, the keyword REF can optionally be preceded by "BRANDED b" where b is a text constant called the brand. Brands distinguish types that would otherwise be the same; they have no other semantic effect. All brands in a program must be distinct. If BRANDED is present and b is absent, the implementation automatically supplies a unique value for b. Explicit brands are useful for persistent data storage.

**Comment by ghost** [comment]
Another thing the OP forgot to mention, CM3 has no GUI library. Trestle is a joke!

https://github.com/modula3/cm3/issues/1121

**Comment by thielema** [comment]
On Sun, 8 Jan 2023, tn1997tn wrote:

> Could you find where in the language definition talk about BRANDED OBJECT?

https://modula3.github.io/cm3/reference/refs.html

**Comment by thielema** [comment]
On Sun, 8 Jan 2023, tn1997tn wrote:

> Could you find where in the language definition talk about BRANDED 
> OBJECT?

https://modula3.github.io/cm3/reference/objects.html

**Comment by ghost** [comment]
No information if not these questions I can't know!

https://github.com/modula3/cm3/issues/1085
https://github.com/modula3/cm3/issues/1086

**Comment by ghost** [comment]
No information on the language definition. If you know CM3 from github and don't know the site of it on elegosoft's website, you can't possibly have any ideas about what is quake without this question: https://github.com/modula3/cm3/issues/1070

**Reply by ghost** [reply]
Yes, you will not know what is .ig, .mg and .tmpl file and what they do too. You need to ask: https://github.com/modula3/cm3/issues/1073 https://github.com/modula3/cm3/issues/1074 Where is the description about them? Where do I find the information about them without asking? Very luckly for me someone asked it before.

**Reply by ghost** [reply]
I made a list of useful questions (including submitted issues and discussion threads), but after the account of the people asked the questions flagged by github I lost most if not all the information. I don't want to ask all of these things again!

**Comment by nsgnkhibdk2cls0f** [comment]
@tn1997tn Please stop spamming my inbox. Each reply of you is one email Github sent me.

**Reply by nsgnkhibdk2cls0f** [reply]
Ah I found the Unsubscribe button. Good riddance @tn1997tn !

**Comment by nsgnkhibdk2cls0f** [comment]
@thielema @RodneyBates @VictorMiasnikov My problems with CM3 is there is no interfaces to external libraries available. Very useful libraries like libcurl, libxm2, sdl2, glfw, gtk,... Even interface to the Windows API is very lacking and outdated. My opinion is if you don't have the resources to be good at all fronts (not everyone is backed by big corporations like Oracle and Microsoft isn't it?), then try to be good at only one front and try the best. Please have a look at the BCX: https://www.bcxbasiccoders.com/ It's a BASIC to C translator for Windows. The author used the word translator but not compiler, he is humble. Because I think CM3 is the same, translating to C and uses C compiler to compile. So CM3 is a translator too, not a compiler. Please have a look at the documentation of BCX: https://bcxbasiccoders.com/webhelp/BCXHelp.htm You could see it's a very complete product that cover most if not all the features you should expect from a commercial product. Supports for Windows programming is excellent with full COM/ActiveX support. The BASIC language is not as good as Modula-3, but BCX provided you the tools to get the job done. What CM3 could give me? An empty language with no tools? I'm not writing an operating system from scratch. I'm not developing web services like Google with Go either. I don't need a good or compact language. Many people is like me.

**Reply by nsgnkhibdk2cls0f** [reply]
Even a toy language like SmallBASIC is more fun to work with and has more things to play with than CM3!

https://smallbasic.github.io/

**Reply by ghost** [reply]
Are you so sure about BCX? I tried to compile some simple programs. It turned out Bc.Exe doesn't check the syntax of the input .bas file. It will produce an equivalent .c file regardless if there is errors in the input .bas file or not. Of course, the users will only know there is problems when they tried to compile the generated .c file. CM3 does check the input for syntax errors. CM3 is a compiler, BCX is really only a translator.

**Reply by nsgnkhibdk2cls0f** [reply]
The error messages of CM3 is the worst I have ever seen. The error messages provide no insights to what's going on. You have to read the interface definition many times to find out what's wrong with the code. What's the point of being a compiled language? The compiler catches errors for you. CM3 is very useless regarding this. BCX translated the code to C. The C compiler will be the one catches errors for you. So yes, BCX doesn't check the syntax of the input, it will translate the code to C anyway, but you will have to pass the C compiler before it could give you the final binary!

**Reply by nsgnkhibdk2cls0f** [reply]
One example: what does `actual not assignable to VALUE formal` mean? This error message is as obscure as possible. You can only know what's going on after reading the interface definition many times! The GCC compiler is much better than this, it will tell you immediately you passed the parameter with the wrong type!

It took many issues only for Sqrt and this obscure error message: https://github.com/modula3/cm3/issues/1076 https://github.com/modula3/cm3/issues/1077 https://github.com/modula3/cm3/issues/1079

**Reply by mikanystrom-intel** [reply]
> One example: what does `actual not assignable to VALUE formal` mean? This error message is as obscure as possible. You can only know what's going on after reading the interface definition many times! The GCC compiler is much better than this, it will tell you immediately you passed the parameter with the wrong type!
> 
> It took many issues only for Sqrt and this obscure error message: #1076 #1077 #1079

VALUE is the default mode of parameter passing in Modula-3.  The others are VAR and READONLY.  If you don't mention the mode, it is VALUE.

The reason that it puts it that way is that the type constraints are different from VALUE parameters as compared to VAR and READONLY parameters.

"formal" is an absolutely standard way of talking about procedure arguments.  They are "formal" rather than "actual" because they don't refer to existing objects.  The objects that are passed in are the "actuals".

It's really not that obscure.  At least I don't think so.

The error messages are pretty good, I think, if you are used to the language.  I think you're probably just unfamiliar with the lingo.

One key thing to keep in mind (and I saw this same thing in the discussion about the wrapping of C libraries): Modula-3 does not really pretend to live in a C universe.  It's from an older, mostly European, tradition (going back to Algol58---M3 is really the most modern language of the Algol family) with its own ideas and practices.  C is newer, from Bell Labs and has a lot of influences from assembly language and generally the more American historical situation of more and better hardware and less recourse to formalities and theoretical thinking.

**Comment by RodneyBates** [comment]
On 1/8/23 11:16, tn1997tn wrote:
>
> REVEAL keyword. It's not documented. People has to learn why there is the REVEAL keyword in books.
>

In Modula-3: Language definition,

2.4.7 Primary description
2.5.4 An example
2.5.5 Another
2.8.1 The keyword
2.8.7 Syntax summary

**Reply by ghost** [reply]
You descended from Modula-2, so learn how to write a proper language report like Prof. Wirth! :smile: 

I can't find HTML version of Modula-2 report, only an ugly scanned PDF. I will give you Component Pascal language report instead :smile: 

https://github.com/pahihu/gpcp-JVM/blob/main/documentation/CP-Lang.pdf

**Reply by ghost** [reply]
I only see a mess in your language definition.

**Reply by ghost** [reply]
Oberon-2 report: https://people.inf.ethz.ch/wirth/Oberon/Oberon.Report.pdf

**Reply by ghost** [reply]
Oberon-07 report: https://miasap.se/obnc/oberon-report.html

**Reply by ghost** [reply]
Look at them and see how bad your language definition is :smile:

**Reply by ghost** [reply]
Your language definition was organized a very illogical way, making it very hard to find the thing someone needs. If it's a whole page HTML like the Oberon-07 report, it's a bit easier to deal with. Scroll from the top and to bottom and you will find what you want regardless of where it was put. The current format is, you have to jump back and forth between sections, have no idea where the thing you wanted to search for located, because it was organized very arbitrary.

**Comment by ghost** [comment]
@thielema I want more details about EXTERNAL than this:

http://modula3.github.io/cm3/reference/complete/html/2_8_5Pragmas.html

Thank you.

**Comment by RodneyBates** [comment]
On 1/8/23 11:22, thielema wrote:
>
> On Sun, 8 Jan 2023, tn1997tn wrote:
>
> > REVEAL keyword. It's not documented. People has to learn why there is the REVEAL keyword in books

The language definition is available in multiple forms.
.

The complete reference in single files is in cm3/master/docs/reference/complete:

   m3-defn-complete.html
   m3-defn-complete.pdf
   m3-defn-complete.tex  Source code.

Between my web browser and text editor, I can search all 3 of these.

The individual chapters in html form are also in
cm3/master/docs/reference.  Even the chapter name
revelations.html is at least a clue.

But the .ps appears to need regeneration.

**Reply by ghost** [reply]
Thank you. CM3 is a big distribution, I can't possibly know all of this. If not your comment I still think it's only available online on http://modula3.github.io/cm3/reference/complete/html/Modula_3_Language_definitio.html

**Comment by RodneyBates** [comment]
On 1/8/23 11:35, tn1997tn wrote:
>
> Only to show your language definition is organized in a way that more difficult for people to find things. When I expect to find it on one section it's indeed on another section that has nothing to do with it. e.g: PRAGMA, fragma is a statement, it should be on Statements, why it's on Syntax?
>
  2.8.5, section title "Pragmas".  In Modula-3, these are not denoted
by "PRAGMA", but by <* ... *>, as the first sentence says. Nor are
they statements, and nor are they part of the language, as it also says.

But even in the language definition, it "recommends" pragma EXTERNAL.

For implementation matters outside the language, see the html
files in cm3/masters/docs/cm3.  File pragmas.html describes
pragmas implemented by CM3.  There's also a quake.html.

In doc/help/cm3/m3build is more implementation info,
including in progs.html, in the m3makefile:

*"build_standalone() *ensures that the program being built does not depend on dynamic linking and shared libraries. To have an effect, this procedure must be called before program or Program is called."

which you asked about.

**Reply by ghost** [reply]
Thank you for being patient with me!

**Reply by ghost** [reply]
What does <* PRAGMA UNALIGNED *> mean? pragmas.html doesn't list it. I found it on WinNT.i3 (m3core). It should have a special meaning for the compiler as the name of it suggests.

**Reply by ghost** [reply]
There are many PRAGMA not listed on pragmas.html. e.g: <* WINAPI *> on WinBase.i3

Does the document simply missed them or they are only _a kind of comment_ that _annotated the code_ which _only has meanings for the programmer_ and _will be ignored by the compiler_?

**Reply by ghost** [reply]
I found CM3's PRAGMA is more like Java's annotation than the `#pragma` compiler directive of C/C++ and Delphi.

**Comment by RodneyBates** [comment]
On 1/13/23 12:05, nsgnkhibdk2cls0f wrote:
>
> One example: what does |actual not assignable to VALUE formal| mean?
>
"Actual parameter" and "formal parameter", often abbreviated
"actual" and "formal", are very common terms going back at
least as far as ALGOL 60.  An actual parameter is the expression
between the parentheses in a call.  A formal is the mode, name,
and type between the parentheses in the declared signature or
prototype of the procedure or function.

C often calls actual parameters "arguments" and formal parameters
just "parameters".

An example that would produce this message:

(* Declaration: *)
PROCEDURE P(A: INTEGER) = ...
BEGIN
   (* Call: *)
   P ( 123.17 * 2.0 )
END

The type of 123.17*2.0 is REAL, not assignable to INTEGER.
The default mode of a formal, if not explicitly specified is VALUE.

**Comment by dragisha** [comment]
cm3 is an open-source project. Free of charge, and free of any developer’s obligations. People developing it have their own reasons and I am happy to get result of their work. Thank you all guys!

You can use it, you can collaborate on it. Or you just can ignore it, as I am ignoring FreeBasic for example.

You need something? Develop it, or pay somebody to develop it, or just continue searching and find it already developed.

Whatever.

**Comment by dragisha** [comment]
Try both. And share experience.

Open wource at work.

**Comment by ghost** [comment]
I found CM3 has many interesting packages as part of the distribution but there are no documents about them, what they are supposed to be used for and how to use them. The only we have is this index: http://modula3.github.io/cm3/help/gen_html/INDEX.html. You could find the link to that index at the end of the CM3 Interface Index page: http://modula3.github.io/cm3/help/interfaces.html
